packages/components/src/calendar/utils/use-controlled-value.ts

```diff
@@ -0,0 +1,61 @@
+/**
+ * WordPress dependencies
+ */
+/**
+ * External dependencies
+ */
+import { useState, useCallback } from '@wordpress/element';
+
+type Props< T > = {
+	defaultValue?: T;
+	value?: T | null | undefined;
+	onChange?: ( newValue: T, ...args: any[] ) => void;
+};
+
+/**
+ * Handles controlled and uncontrolled state for the selected calendar value.
+ * It is assumed that the `value` prop is controlled when it's not undefined:
+ * - initial date selected, uncontrolled: use a non-undefined`defaultValue`
+ * - initial date selected, controlled: use a non-undefined `value`
+ * - no date selected, controlled: set `value` to `null`
+ *
+ * The `onChange` prop will return `undefined` when no date is selected,
+ * regardless of controlled / uncontrolled. It is expected that the consumer
+ * of the component will handle setting the value to `null` to indicate no date
+ * selected in controlled mode.
+ * @param root0
+ * @param root0.defaultValue
+ * @param root0.onChange
+ * @param root0.value
+ */
+export function useControlledValue< T >( {
```

Comment:
As a follow-up, we should try to re-use the existing `useControlledValue` hook in the package https://github.com/WordPress/gutenberg/blob/2dd93a3b957906412f6b4d8c547d8326f2ba8b78/packages/components/src/calendar/utils/use-controlled-value.ts